### PR TITLE
API-21863 claims use v3 for token validation

### PR DESCRIPTION
**Pull request by crolarlibertyva**
## Summary

- *To prepare for the removal of the OpenidApplicationController, the claims api is modified to use the v3 token validation service deployed in dvp instead of relying on OpenidApplicationController as a superclass*
- *Modification owned by Pivot, though Pivot is not the owner of the claims api*

## Related issue(s)
- *https://vajira.max.gov/browse/API-21863*

## Testing done

- *Tested locally for some cases*
- *Further testing needed on a deployed developer instance*

## Screenshots
NA

## What areas of the site does it impact?
*All claims api endpoints will rely on an external service for token validation*

## Acceptance criteria

- [ ]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [ ]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

**Comment by FonzMP** [comment]
We're getting there, but there's a trickle effect now where we resolve one issue just to find another. Currently it looks like we're throwing a 500 now due to [this line](https://github.com/department-of-veterans-affairs/vets-api/blob/68f0a0d4768c35f93d3c1fff4abe0ab34e69e9fd/modules/claims_api/app/controllers/claims_api/v1/forms/power_of_attorney_controller.rb#LL184C38-L184C38). 

```
"exception": "undefined method `email' for #<ClaimsApi::ClaimsUser.....
```

Also curious, where do we stand on the CSRF protection issue? That's something I haven't been looking in to. Curious if you need eyes on that as well.

**Comment by crolarlibertyva** [comment]
> We're getting there, but there's a trickle effect now where we resolve one issue just to find another. Currently it looks like we're throwing a 500 now due to [this line](https://github.com/department-of-veterans-affairs/vets-api/blob/68f0a0d4768c35f93d3c1fff4abe0ab34e69e9fd/modules/claims_api/app/controllers/claims_api/v1/forms/power_of_attorney_controller.rb#LL184C38-L184C38).
> 
> ```
> "exception": "undefined method `email' for #<ClaimsApi::ClaimsUser.....
> ```
> 
> Also curious, where do we stand on the CSRF protection issue? That's something I haven't been looking in to. Curious if you need eyes on that as well.

Changes put in to provide the email field.

File: modules/claims_api/app/controllers/concerns/claims_api/token_validation.rb [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+      def verify_access!
+        verify_access_token!
+      rescue
+        render_unauthorized
+      end
+
+      # Determine if the current authenticated user is allowed access
+      # raise if current authenticated user is neither the target veteran, nor target veteran representative
+
+      def verify_access_token!
```
Comment (FonzMP) [inline modules/claims_api/app/controllers/concerns/claims_api/token_validation.rb:26]: looks like we're calling this twice in v2. For example: 
Claims Controller has 
`before_action :verify_access!`
Application Controller has
`before_action :authenticate`

I believe v1 is looking good though.

Comment (crolarlibertyva) [reply modules/claims_api/app/controllers/concerns/claims_api/token_validation.rb:26]: Pulled it out of claims controller.

File: modules/claims_api/app/controllers/claims_api/v1/application_controller.rb [hunk 2]
```diff
@@ -8,12 +8,16 @@
...
-    class ApplicationController < ::OpenidApplicationController
+    class ApplicationController < ::ApplicationController
       include ClaimsApi::MPIVerification
       include ClaimsApi::HeaderValidation
       include ClaimsApi::JsonFormatValidation
+      include ClaimsApi::TokenValidation
       include ClaimsApi::CcgTokenValidation
-
+      include ClaimsApi::TargetVeteran
+      skip_before_action :verify_authenticity_token
```
[the remark this answers is not in the document]
Comment (crolarlibertyva) [reply modules/claims_api/app/controllers/claims_api/v1/application_controller.rb:18]: This API requires an access token which is subsequently validated here. Potential CSRF vulnerability is not applicable here.

**Review by dbrown-va — APPROVED** [review]
This will require a dangerbot override, but likely necessary due to the scale of rspec's effected from the change.

Moving from inheritance via controller to calling token validation as a service reduces the interdependency and ultimately allows for extraction to a microservice.

`with_okta_user` -> `mock_acg` is a also required with this change.

Note: Will want to coordinate this merge following the daily vets-api deploy to maximize testing time in Dev/Staging environments ahead of auto-deploys to higher envs.

**Review by ryan-mcneil — APPROVED** [review]
🤞

File: modules/claims_api/app/controllers/concerns/claims_api/target_veteran.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+require 'rest-client'
+
+module ClaimsApi
+  module TargetVeteran
+    extend ActiveSupport::Concern
+
+    included do
+      def target_veteran
+        @target_veteran ||= if @is_valid_ccg_flow
+                              build_target_veteran(veteran_id: params[:veteranId], loa: { current: 3, highest: 3 })
```
Comment (dbrown-va) [inline modules/claims_api/app/controllers/concerns/claims_api/target_veteran.rb:12]: Just a comment for historical tracking, these access tokens are associated with LOA3 accounts only when attended login is completed.

File: modules/claims_api/spec/requests/v1/claims_request_spec.rb [hunk 4]
```diff
@@ -47,7 +48,7 @@
 
   context 'index' do
     it 'lists all Claims', run_at: 'Tue, 12 Dec 2017 03:09:06 GMT' do
-      with_okta_user(scopes) do |auth_header|
```
Comment (dbrown-va) [inline modules/claims_api/spec/requests/v1/claims_request_spec.rb:50]: Really appreciate the adjustment here to decouple from okta.

Drives us towards the ultimate objective from a security standpoint.

**Review by dbrown-va — APPROVED** [review]
Moves to calling token validation as a service rather than the parent inheritance method.  Significantly decouples complexity between validation and API layer components.

**Review by FonzMP — APPROVED** [review]
Pending the production settings values, I think it's good to go. :shipit:

**Pull request merged**